libraries/ts/src/margin/serum/market.ts

```diff
@@ -1,785 +1,421 @@
 import assert from "assert"
-import { blob, struct, u8 } from "buffer-layout"
-import { BN } from "@project-serum/anchor"
 import {
-  decodeEventQueue,
-  decodeRequestQueue,
-  getFeeTier,
-  getLayoutVersion,
-  Market as SerumMarket,
-  OpenOrders,
-  supportsSrmFeeDiscounts
-} from "@project-serum/serum"
-import { getMintDecimals, MarketOptions, Order, ORDERBOOK_LAYOUT, OrderParams } from "@project-serum/serum/lib/market"
-import { Slab } from "@project-serum/serum/lib/slab"
-import {
-  closeAccount,
-  initializeAccount,
-  MSRM_DECIMALS,
-  MSRM_MINT,
-  SRM_DECIMALS,
-  SRM_MINT,
-  TOKEN_PROGRAM_ID,
-  WRAPPED_SOL_MINT
-} from "@project-serum/serum/lib/token-instructions"
-import {
-  Account,
-  AccountInfo,
-  Commitment,
-  Connection,
-  LAMPORTS_PER_SOL,
+  Keypair,
   PublicKey,
   SystemProgram,
   SYSVAR_RENT_PUBKEY,
   Transaction,
   TransactionInstruction,
-  TransactionSignature
+  Signer
 } from "@solana/web3.js"
-
+import { AnchorProvider, BN } from "@project-serum/anchor"
+import { getLayoutVersion, Market as SerumMarket, Orderbook as SerumOrderbook, OpenOrders } from "@project-serum/serum"
+import { MarketOptions, Order } from "@project-serum/serum/lib/market"
+import {
+  closeAccount,
+  initializeAccount,
+  TOKEN_PROGRAM_ID,
+  WRAPPED_SOL_MINT
+} from "@project-serum/serum/lib/token-instructions"
+import { TokenAmount } from "src/token"
+import { MarginMarketConfig, MarginMarkets, MarginTokens } from "../config"
+import { PoolTokenChange } from "../pool"
 import { MarginAccount } from "../marginAccount"
 import { MarginPrograms } from "../marginClient"
-import { Number192 } from "../../utils"
 
+export type selfTradeBehavior = "decrementTake" | "cancelProvide" | "abortTransaction"
+export type orderSide = "sell" | "buy" | "ask" | "bid"
+export type orderType = "limit" | "ioc" | "postOnly"
 export class Market {
-  private _programs: MarginPrograms
-  _decoded: any
-  private _baseSplTokenDecimals: number
-  private _quoteSplTokenDecimals: number
-  private _skipPreflight: boolean
-  private _commitment: Commitment
-  private _serumProgramId: PublicKey
-  private _openOrdersAccountsCache: {
-    [publickKey: string]: { accounts: OpenOrders[]; ts: number }
-  }
-  private _layoutOverride?: any
-
-  private _feeDiscountKeysCache: {
-    [publicKey: string]: {
-      accounts: Array<{
-        balance: number
-        mint: PublicKey
-        pubkey: PublicKey
-        feeTier: number
-      }>
-      ts: number
-    }
-  }
-
-  private marginSerumProgramId: PublicKey
-  private marginSerumAdapterMetadata: PublicKey
+  provider: AnchorProvider
+  programs: MarginPrograms
+  marketConfig: MarginMarketConfig
+  serum: SerumMarket
```

Comment:
All these fields are good. I don't think the market deserves a provider because it's not tied to a specific user. MarginAccount get's a provider because the provider wallet literally owns the margin account and can sign to change it
```suggestion
  programs: MarginPrograms
  marketConfig: MarginMarketConfig
  serum: SerumMarket
```